generator/test_case_generator.go

```diff
@@ -60,35 +70,30 @@ func genMatrix(targetOS string, testDirList []string) []map[string]string {
 	var testMatrix []map[string]string
 	err = json.Unmarshal(byteValueTestMatrix, &testMatrix)
 	if err != nil {
-		log.Panicf("can't unmarshall file %v_test_matrix.json err %v", targetOS, err)
+		log.Panicf("can't unmarshall file %v_test_matrix.json err %v", testType, err)
 	}
 
-	var testMatrixComplete []map[string]string
+	testMatrixComplete := make([]matrixLine, 0, len(testMatrix))
 	for _, test := range testMatrix {
 		for _, testDirectory := range testDirList {
-			testLine := copyMap(test)
-			testLine[testDir] = testDirectory
-			testMatrixComplete = append(testMatrixComplete, testLine)
+			matrixLine := matrixLine{TestDir: testDirectory, TestType: testType}
+			err = mapstructure.Decode(test, &matrixLine)
```

Comment:
nit: seems like the local variable `matrixLine` shadows the struct name `matrixLine`. Consider renaming it.